src/txSerializers.ts

```diff
@@ -77,36 +80,47 @@ const serializeTxCertificate = (certificate: Certificate) => {
     }
 }
 
+const serializeCollateral = (collateral: Collateral) =>
+    [collateral.transactionId, collateral.index]
+
 export const serializeTxBody = (txBody: TransactionBody) => new Map(([
-    [0, txBody.inputs.map(serializeTxInput)],
-    [1, txBody.outputs.map(serializeTxOutput)],
-    [2, identity(txBody.fee)],
-    [3, identity(txBody.ttl)],
-    [4, txBody.certificates?.map(serializeTxCertificate)],
-    [5, txBody.withdrawals && serializeWithdrawals(txBody.withdrawals)],
-    [6, identity(txBody.update)],
-    [7, identity(txBody.metadataHash)],
-    [8, identity(txBody.validityIntervalStart)],
-    [9, txBody.mint && serializeMultiasset(txBody.mint)],
+    [TransactionBodyKeys.INPUTS, txBody.inputs.map(serializeTxInput)],
+    [TransactionBodyKeys.OUTPUTS, txBody.outputs.map(serializeTxOutput)],
+    [TransactionBodyKeys.FEE, identity(txBody.fee)],
+    [TransactionBodyKeys.TTL, identity(txBody.ttl)],
+    [TransactionBodyKeys.CERTIFICATES, txBody.certificates?.map(serializeTxCertificate)],
+    [TransactionBodyKeys.WITHDRAWALS, txBody.withdrawals && serializeWithdrawals(txBody.withdrawals)],
+    [TransactionBodyKeys.UPDATE, identity(txBody.update)],
+    [TransactionBodyKeys.METADATA_HASH, identity(txBody.metadataHash)],
+    [TransactionBodyKeys.VALIDITY_INTERVAL_START, identity(txBody.validityIntervalStart)],
+    [TransactionBodyKeys.MINT, txBody.mint && serializeMultiasset(txBody.mint)],
+    [TransactionBodyKeys.SCRIPT_DATA_HASH, txBody.scriptDataHash],
+    [TransactionBodyKeys.COLLATERAL_INPUTS, txBody.collaterals?.map(serializeCollateral)],
+    [TransactionBodyKeys.REQUIRED_SIGNERS, txBody.requiredSigners],
+    [TransactionBodyKeys.NETWORK_ID, txBody.networkId],
 ]).filter(([_, value]) => value !== undefined) as [number, unknown][])
 
-export const serializeTx = (tx: Transaction) => [
-    serializeTxBody(tx.body),
-    tx.witnessSet,
-    tx.auxiliaryData,
-]
+export const serializeTx = (tx: Transaction) => {
+    return [
+        serializeTxBody(tx.body),
+        tx.witnessSet,
+        tx.scriptValidity,
+        tx.auxiliaryData,
+    ].filter((item) => item !== undefined)
+}
 
 export const serializeRawTx = (rawTx: RawTransaction) => {
-    // older versions of cardano-cli did not include scriptWitnesses
-    if (rawTx.scriptWitnesses === undefined) {
-        return [
-            serializeTxBody(rawTx.body),
-            rawTx.auxiliaryData,
-        ]
+    if (rawTx.scriptWitnesses !== undefined) {
+        // cardano-cli expects indefinite-length scriptWitnesses
+        addIndefiniteLengthFlag(rawTx.scriptWitnesses)
     }
+
     return [
         serializeTxBody(rawTx.body),
         rawTx.scriptWitnesses,
+        rawTx.datumWitnesses,
+        rawTx.redeemerWitnesses,
+        rawTx.scriptValidity,
         rawTx.auxiliaryData,
-    ]
+    ].filter((item) => item !== undefined)
```

Comment:
Again suspicious, how is the caller of this supposed to know which item is which if they don't have fixed positions?

Reply:
I don't think there's much we can do about this - cardano-cli is the source of this ugliness and the result of this function is used only as an input for CBOR encoding. "Items which don't have fixed positions" is what the client expects of us when he wants us to return a CBOR-encoded tx. Or am I missing something?

Reply:
Ah, it's the internal cardano-cli format? So that we are basically returning what arrived and don't care what will be done with it.  Btw. the link at RawTransaction returns error 404. 
Anyway, we will remove the support, so we don't need to dwell on polishing this.